src/typeEvaluator/optimizations.ts

```diff
@@ -8,53 +8,71 @@ function typeNodesSorter(a: TypeNode, b: TypeNode): number {
   return compare(hashField(a), hashField(b))
 }
 
+const hashCache = new WeakMap<TypeNode, string>()
+
 export function hashField(field: TypeNode): string {
```

Comment:
How about moving the caching into a wrapper function instead?

```ts
const hashCache = new WeakMap<TypeNode, string>()

export function hashField(field: TypeNode): string {
  if (hashCache.has(field)) {
    return hashCache.get(field)
  }

  const res = hashFieldUncached(field)
  hashCache.set(field, res)
  return res
}

export function hashFieldUncached(field: TypeNode): string {
 // Same code (which calls `hashField`)
}
```

Reply:
Done!